### PR TITLE
Fix changing column from int to bigint with autoincrement

### PR DESCRIPTION
**Pull request by icewind1991**
Currently it tries to alter the column type to `BIGSERIAL` which is not a valid type to be used in an `ALTER TABLE` query.

Also it currently drops the default which breaks the auto increment.

cc @nickvergessen

### CODE REVIEW

File: lib/Doctrine/DBAL/Platforms/PostgreSqlPlatform.php [hunk 1]
```diff
@@ -534,12 +535,26 @@ public function getAlterTableSQL(TableDiff $diff)
...
+
                 // here was a server version check before, but DBAL API does not support this anymore.
-                $query = 'ALTER ' . $oldColumnName . ' TYPE ' . $type->getSQLDeclaration($column->toArray(), $this);
+                $query = 'ALTER ' . $oldColumnName . ' TYPE ' . $type->getSQLDeclaration($columnDefinition, $this);
                 $sql[] = 'ALTER TABLE ' . $diff->getName($this)->getQuotedName($this) . ' ' . $query;
             }
 
-            if ($columnDiff->hasChanged('default') || $columnDiff->hasChanged('type')) {
+            if ($columnDiff->fromColumn) {
+                $oldTypeIsNumeric = $columnDiff->fromColumn->getType()->getName() === Type::INTEGER || $columnDiff->fromColumn->getType()->getName() === Type::BIGINT;
```
Comment (lcobucci) [inline lib/Doctrine/DBAL/Platforms/PostgreSqlPlatform.php]: Please don't rely on `Type::getName()`, use `instance of` instead.

File: lib/Doctrine/DBAL/Platforms/PostgreSqlPlatform.php [hunk 2]
```diff
@@ -534,12 +535,26 @@ public function getAlterTableSQL(TableDiff $diff)
...
+                $newTypeIsNumeric = $column->getType()->getName() === Type::INTEGER || $column->getType()->getName() === Type::BIGINT;
```
Comment (lcobucci) [inline lib/Doctrine/DBAL/Platforms/PostgreSqlPlatform.php]: Please don't rely on `Type::getName()`, use `instance of` instead.

File: tests/Doctrine/Tests/DBAL/Functional/Schema/PostgreSqlSchemaManagerTest.php [hunk 3]
```diff
@@ -419,6 +419,29 @@ public function testListNegativeColumnDefaultValue()
         self::assertEquals(-1.1, $columns['col_decimal']->getDefault());
         self::assertEquals('(-1)', $columns['col_string']->getDefault());
     }
+
+    public function testAlterTableAutoIncrementBigInt()
```
Comment (lcobucci) [inline tests/Doctrine/Tests/DBAL/Functional/Schema/PostgreSqlSchemaManagerTest.php]: ` : void` missing

File: tests/Doctrine/Tests/DBAL/Functional/Schema/PostgreSqlSchemaManagerTest.php [hunk 4]
```diff
@@ -419,6 +419,29 @@ public function testListNegativeColumnDefaultValue()
...
+    {
+        $tableFrom = new \Doctrine\DBAL\Schema\Table('autoinc_table_bitgint');
```
Comment (lcobucci) [inline tests/Doctrine/Tests/DBAL/Functional/Schema/PostgreSqlSchemaManagerTest.php]: Please simplify the FQCN

File: tests/Doctrine/Tests/DBAL/Functional/Schema/PostgreSqlSchemaManagerTest.php [hunk 5]
```diff
@@ -419,6 +419,29 @@ public function testListNegativeColumnDefaultValue()
...
+        $column = $tableFrom->addColumn('id', 'integer');
+        $column->setAutoincrement(true);
+        $this->_sm->createTable($tableFrom);
+        $tableFrom = $this->_sm->listTableDetails('autoinc_table_bitgint');
+        self::assertTrue($tableFrom->getColumn('id')->getAutoincrement());
+
+        $tableTo = new \Doctrine\DBAL\Schema\Table('autoinc_table_bitgint');
```
Comment (lcobucci) [inline tests/Doctrine/Tests/DBAL/Functional/Schema/PostgreSqlSchemaManagerTest.php]: Please simplify the FQCN

File: tests/Doctrine/Tests/DBAL/Functional/Schema/PostgreSqlSchemaManagerTest.php [hunk 6]
```diff
@@ -419,6 +419,29 @@ public function testListNegativeColumnDefaultValue()
...
+        $column = $tableTo->addColumn('id', 'bigint');
+        $column->setAutoincrement(true);
+
+        $c = new \Doctrine\DBAL\Schema\Comparator();
```
Comment (lcobucci) [inline tests/Doctrine/Tests/DBAL/Functional/Schema/PostgreSqlSchemaManagerTest.php]: Please simplify the FQCN

File: tests/Doctrine/Tests/DBAL/Functional/Schema/PostgreSqlSchemaManagerTest.php [hunk 7]
```diff
@@ -419,6 +419,29 @@ public function testListNegativeColumnDefaultValue()
...
+        $diff = $c->diffTable($tableFrom, $tableTo);
+        self::assertInstanceOf('Doctrine\DBAL\Schema\TableDiff', $diff, "There should be a difference and not false being returned from the table comparison");
+        self::assertEquals(array("ALTER TABLE autoinc_table_bitgint ALTER id TYPE BIGINT"), $this->_conn->getDatabasePlatform()->getAlterTableSQL($diff));
```
Comment (lcobucci) [inline tests/Doctrine/Tests/DBAL/Functional/Schema/PostgreSqlSchemaManagerTest.php]: `assertSame()` is more appropriated here

Comment (lcobucci) [reply tests/Doctrine/Tests/DBAL/Functional/Schema/PostgreSqlSchemaManagerTest.php]: Please also use short-array syntax

File: tests/Doctrine/Tests/DBAL/Functional/Schema/PostgreSqlSchemaManagerTest.php (hunk 3 above)
Comment (lcobucci) [reply tests/Doctrine/Tests/DBAL/Functional/Schema/PostgreSqlSchemaManagerTest.php]: Please add `@group 2916` to a docblock here

**Review by lcobucci — CHANGES_REQUESTED** [review]
Please rebase your branch to sync to the latest changes of `master`

**Comment by icewind1991** [comment]
@lcobucci all done

File: tests/Doctrine/Tests/DBAL/Functional/Schema/PostgreSqlSchemaManagerTest.php [hunk 8]
```diff
@@ -463,6 +465,32 @@ public function testAutoIncrementCreatesSerialDataTypesWithoutADefaultValueEvenW
...
+        $tableFrom = $this->_sm->listTableDetails('autoinc_table_bitgint');
+        self::assertTrue($tableFrom->getColumn('id')->getAutoincrement());
+
+        $tableTo = new Table('autoinc_table_bitgint');
+        $column = $tableTo->addColumn('id', 'bigint');
+        $column->setAutoincrement(true);
+
+        $c = new Comparator();
+        $diff = $c->diffTable($tableFrom, $tableTo);
+        self::assertInstanceOf('Doctrine\DBAL\Schema\TableDiff', $diff, "There should be a difference and not false being returned from the table comparison");
```
Comment (lcobucci) [inline tests/Doctrine/Tests/DBAL/Functional/Schema/PostgreSqlSchemaManagerTest.php]: Please use `TableDiff::class` instead

File: lib/Doctrine/DBAL/Platforms/PostgreSqlPlatform.php [hunk 9]
```diff
@@ -536,12 +537,26 @@ public function getAlterTableSQL(TableDiff $diff)
...
+                $columnDefinition['autoincrement'] = false;
+
                 // here was a server version check before, but DBAL API does not support this anymore.
-                $query = 'ALTER ' . $oldColumnName . ' TYPE ' . $type->getSQLDeclaration($column->toArray(), $this);
+                $query = 'ALTER ' . $oldColumnName . ' TYPE ' . $type->getSQLDeclaration($columnDefinition, $this);
                 $sql[] = 'ALTER TABLE ' . $diff->getName($this)->getQuotedName($this) . ' ' . $query;
             }
 
-            if ($columnDiff->hasChanged('default') || $columnDiff->hasChanged('type')) {
+            if ($columnDiff->fromColumn) {
```
Comment (lcobucci) [inline lib/Doctrine/DBAL/Platforms/PostgreSqlPlatform.php]: What do you think about extracting this `if` block to a private method? I think that something like would improve the readability a bit:

```php
private function hasDefaultTypeChanged(ColumnDiff $columnDiff, Column $newColumn) : bool
{
    $defaultTypeChanged = $columnDiff->hasChanged('type');

    if ( ! $columnDiff->fromColumn || ! $defaultTypeChanged) {
        return $defaultTypeChanged;
    }

    $oldTypeIsNumeric = $this->isNumericType($columnDiff->fromColumn->getType());
    $newTypeIsNumeric = $this->isNumericType($newColumn->getType());

    return ! ($oldTypeIsNumeric && $newTypeIsNumeric && $newColumn->getDefault() === null)
}
```

We might improve things there (specially naming wise), but I think it makes things easier and more understandable 😄

**Review by lcobucci — CHANGES_REQUESTED** [review]
Nice job, some more nitpicks 😜

~Just to understand, is this needed only for PGSQL 10+ compatibility? I mean, I didn't see any issue reporting this before.~
~This question is just to categorise this PR correctly.~

I've already checked, this operation breaks badly on PGSQL 9.x too

**Comment by lcobucci** [comment]
One last request, could add some info on the commit msg on why this is needed? BTW thanks for ensuring that the commit history is organised =)

**Comment by lcobucci** [comment]
Ahhh I forgot one more thing, can you please add a test for the reverse operation too? bigint -> int

**Comment by icewind1991** [comment]
done

**Review by lcobucci — APPROVED** [review]

**Comment by lcobucci** [comment]
@icewind1991 🚢 and backported to `2.6.x` (with some minor improvements). Thanks for your contribution!

**Comment by icewind1991** [comment]
Any chance of getting things backported to 2.5.x? We're stuck with supporting php 5.6/7.0 for the foreseeable future

**Pull request merged**